# backtesting

A quick test of the portfolio methodology via backtesting, with some helpful functions. 

# Setup

## Imports
Useful dependencies.

In [2]:
import pandas as pd
import quandl as ql
import numpy as np
import matplotlib.pyplot as plt
import cvxopt as opt
from cvxopt import blas, solvers
import warnings

## Options
Convenience.

In [3]:
#%matplotlib inline                       # Make charts bigger
pd.options.display.max_rows = 4000       # Show 4000 table rows?
np.set_printoptions(threshold=np.inf)    # Scroll to infinity
solvers.options['show_progress'] = False # Less busy convex solver logs

## Data Retrieval

Retrieving daily altcoin data from [Kaggle.](https://www.kaggle.com/jessevent/all-crypto-currencies/data) Would prefer Quandl, but costs [💰](https://www.quandl.com/data/BNC2-BNC-Digital-Currency-Indexed-EOD/pricing?license=282&plan=596&step=2).

In [4]:
df = pd.read_csv('data/crypto.csv')

In [5]:
df_tickers = df.symbol.unique()
print(df_tickers)

['BTC' 'XRP' 'ETH' 'BCH' 'ADA' 'LTC' 'XLM' 'XEM' 'MIOTA' 'DASH' 'XMR' 'NEO'
 'TRX' 'EOS' 'BTG' 'QTUM' 'XRB' 'ETC' 'BCC' 'ICX' 'LSK' 'BTS' 'XVG' 'OMG'
 'SNT' 'ARDR' 'STEEM' 'ZEC' 'PPT' 'STRAT' 'USDT' 'WAVES' 'BCN' 'HSR' 'DOGE'
 'KMD' 'SC' 'GNT' 'BNB' 'REP' 'VEN' 'VERI' 'ARK' 'KCS' 'DCR' 'DGB' 'SALT'
 'PIVX' 'DRGN' 'NXT' 'FCT' 'AION' 'MONA' 'GBYTE' 'BAT' 'MAID' 'REQ' 'XZC'
 'ENG' 'POWR' 'BTCD' 'CVC' 'ZRX' 'RDD' 'PAY' 'XP' 'ETN' 'KNC' 'WAX' 'SYS'
 'SAN' 'FUN' 'ETHOS' 'BTM' 'RHOC' 'DGD' 'QASH' 'KIN' 'XBY' 'AE' 'GAS' 'VTC'
 'GNO' 'GAME' 'STORJ' 'ICN' 'QSP' 'DENT' 'SUB' 'ELF' 'GXS' 'SKY' 'UBQ'
 'RDN' 'WTC' 'CTR' 'MANA' 'BNT' 'STORM' 'BAY' 'BLOCK' 'LINK' 'EDG' 'SMART'
 'POE' 'RPX' 'TNB' 'ITC' 'TRIG' 'R' 'NAV' 'CNX' 'MCO' 'EMC2' 'BCO' 'RCN'
 'LEND' 'NXS' 'DATA' 'MTL' 'CMT' 'ADX' 'SNGLS' 'ZCL' 'CLOAK' 'ANT' 'PART'
 'PPP' 'RLC' 'DCN' 'PRE' 'ZEN' 'CND' 'PPC' 'UTK' 'RISE' 'XDN' 'PAC' 'SNM'
 'ENJ' 'LRC' 'EMC' 'XPA' 'VEE' 'TNT' 'THC' 'AGRS' 'QRL' 'MED' 'LBC' 'VOX'
 'SRN' 'DBC' 'DCT' 'GRS' 'MGO' 'WAB

In [58]:
# Test
df.loc[df.symbol == 'XRP']['date'][:1].max()
df.loc[df.symbol == 'XRP']['name'][:1].max()

'Ripple'

In [ ]:
# Wrap

def getContext(available_tickers):
    checked_tickers = []
    
    for ticker in available_tickers:
        # Name
        name = df.loc[df.symbol == ticker]['name'][:1].max()
        
        # Date
        firstDate = df.loc[df.symbol == ticker]['date'][:1].max()
        lastDate = df.loc[df.symbol == ticker]['date'][-1:].max()
       
        # Market Cap
        firstMarketCap = df.loc[df.symbol == ticker]['market'][:1].max()
        lastMarketCap = df.loc[df.symbol == ticker]['market'][-1:].max()
        
        # Open
        firstOpenPrice = df.loc[df.symbol == ticker]['open'][:1].max()
        lastOpenPrice = df.loc[df.symbol == ticker]['open'][-1:].max()
        
        # Log
        print('{} ({}) \n Data available from {} to {}.\n Market Cap grew from ${} to ${}.\n Price grew from (${} to ${})\n'.format(ticker, name, firstDate, lastDate, firstMarketCap, lastMarketCap, firstOpenPrice, lastOpenPrice))
        
        # Append
        checked_tickers.append((ticker, name, firstDate, lastDate, firstMarketCap, lastMarketCap, firstOpenPrice, lastOpenPrice))
        
    return checked_tickers
    
#print(df_tickers)    
getContext(df_tickers[1:100])

XRP (Ripple) 
 Data available from 2013-08-04 to 2018-01-02.
 Market Cap grew from $45921000 to $91791800000.
 Price grew from ($0.005874 to $2.37)

ETH (Ethereum) 
 Data available from 2015-08-07 to 2018-01-02.
 Market Cap grew from $0 to $74695700000.
 Price grew from ($2.83 to $772.35)

BCH (Bitcoin Cash) 
 Data available from 2017-07-23 to 2018-01-02.
 Market Cap grew from $0 to $41114800000.
 Price grew from ($555.89 to $2434.5)

ADA (Cardano) 
 Data available from 2017-10-01 to 2018-01-02.
 Market Cap grew from $0 to $18788700000.
 Price grew from ($0.021678 to $0.724676)

LTC (Litecoin) 
 Data available from 2013-04-28 to 2018-01-02.
 Market Cap grew from $73773400 to $12496900000.
 Price grew from ($4.3 to $228.99)

XLM (Stellar) 
 Data available from 2014-08-05 to 2018-01-02.
 Market Cap grew from $772597 to $8873700000.
 Price grew from ($0.0029760000000000003 to $0.49687600000000004)

XEM (NEM) 
 Data available from 2015-04-01 to 2018-01-02.
 Market Cap grew from $3604110 to

In [ ]:
def makeBenchmarks(available_tickers):
    # All Time Low
    firstMarketCap = df.loc[df.symbol == ticker]['market'][1:]
    lastMarketCap = df.loc[df.symbol == ticker]['market'][:-1]

def makeReturns(available_tickers):
        VWAP_simple =  df.loc[df.symbol == ticker][''][:-1]

In [45]:
df_coins

array([['BTC', 'bitcoin'],
       ['XRP', 'ripple'],
       ['ETH', 'ethereum'],
       ..., 
       ['ADK', '10mtoken'],
       ['CSC', 'infinity-pay'],
       ['BTU', 'darklisk']],
      dtype='<U31')

In [22]:
df.loc[df.symbol == 'STEEM'].tail(100)

slug symbol   name        date  ranknow      open     high       low  \
18093  steem  STEEM  Steem  2017-09-25       27  1.070000  1.17000  1.040000   
18094  steem  STEEM  Steem  2017-09-26       27  1.110000  1.12000  1.070000   
18095  steem  STEEM  Steem  2017-09-27       27  1.080000  1.26000  1.080000   
18096  steem  STEEM  Steem  2017-09-28       27  1.220000  1.49000  1.200000   
18097  steem  STEEM  Steem  2017-09-29       27  1.440000  1.64000  1.260000   
18098  steem  STEEM  Steem  2017-09-30       27  1.640000  1.68000  1.430000   
18099  steem  STEEM  Steem  2017-10-01       27  1.460000  1.52000  1.410000   
18100  steem  STEEM  Steem  2017-10-02       27  1.390000  1.43000  1.290000   
18101  steem  STEEM  Steem  2017-10-03       27  1.320000  1.33000  1.170000   
18102  steem  STEEM  Steem  2017-10-04       27  1.220000  1.23000  1.130000   
18103  steem  STEEM  Steem  2017-10-05       27  1.180000  1.22000  1.090000   
18104  steem  STEEM  Steem  2017-10-06       27  1.170000  1.23000  1.170000   
18105  steem  STEEM  Steem  2017-10-07       27  1.210000  1.26000  1.170000   
18106  steem  STEEM  Steem  2017-10-08       27  1.220000  1.27000  1.080000   
18107  steem  STEEM  Steem  2017-10-09       27  1.120000  1.15000  0.958003   
18108  steem  STEEM  Steem  2017-10-10       27  1.070000  1.13000  1.020000   
18109  steem  STEEM  Steem  2017-10-11       27  1.030000  1.12000  1.030000   
18110  steem  STEEM  Steem  2017-10-12       27  1.100000  1.12000  1.010000   
18111  steem  STEEM  Steem  2017-10-13       27  1.040000  1.04000  0.967398   
18112  steem  STEEM  Steem  2017-10-14       27  1.030000  1.18000  1.010000   
18113  steem  STEEM  Steem  2017-10-15       27  1.070000  1.10000  0.998666   
18114  steem  STEEM  Steem  2017-10-16       27  1.070000  1.08000  1.010000   
18115  steem  STEEM  Steem  2017-10-17       27  1.050000  1.09000  1.010000   
18116  steem  STEEM  Steem  2017-10-18       27  1.060000  1.11000  0.971589   
18117  steem  STEEM  Steem  2017-10-19       27  1.070000  1.09000  1.000000   
18118  steem  STEEM  Steem  2017-10-20       27  1.090000  1.13000  0.993585   
18119  steem  STEEM  Steem  2017-10-21       27  0.999598  1.02000  0.893548   
18120  steem  STEEM  Steem  2017-10-22       27  0.972041  1.03000  0.939994   
18121  steem  STEEM  Steem  2017-10-23       27  0.974869  0.98807  0.880831   
18122  steem  STEEM  Steem  2017-10-24       27  0.930480  1.07000  0.898789   
...      ...    ...    ...         ...      ...       ...      ...       ...   
18163  steem  STEEM  Steem  2017-12-04       27  1.220000  1.46000  1.190000   
18164  steem  STEEM  Steem  2017-12-05       27  1.430000  1.62000  1.420000   
18165  steem  STEEM  Steem  2017-12-06       27  1.450000  1.48000  1.260000   
18166  steem  STEEM  Steem  2017-12-07       27  1.290000  2.15000  1.270000   
18167  steem  STEEM  Steem  2017-12-08       27  2.140000  2.16000  1.610000   
18168  steem  STEEM  Steem  2017-12-09       27  1.900000  1.94000  1.540000   
18169  steem  STEEM  Steem  2017-12-10       27  1.670000  1.92000  1.310000   
18170  steem  STEEM  Steem  2017-12-11       27  1.680000  1.79000  1.680000   
18171  steem  STEEM  Steem  2017-12-12       27  1.720000  1.87000  1.660000   
18172  steem  STEEM  Steem  2017-12-13       27  1.860000  1.88000  1.690000   
18173  steem  STEEM  Steem  2017-12-14       27  1.760000  2.10000  1.760000   
18174  steem  STEEM  Steem  2017-12-15       27  2.040000  2.08000  1.800000   
18175  steem  STEEM  Steem  2017-12-16       27  1.970000  2.26000  1.970000   
18176  steem  STEEM  Steem  2017-12-17       27  2.110000  2.42000  2.100000   
18177  steem  STEEM  Steem  2017-12-18       27  2.330000  2.94000  2.210000   
18178  steem  STEEM  Steem  2017-12-19       27  2.870000  3.37000  2.630000   
18179  steem  STEEM  Steem  2017-12-20       27  2.880000  3.70000  2.430000   
18180  steem  STEEM  Steem  2017-12-21       27  3.440000  4.17000  3.310000   
18181  s

17568    Steem
17569    Steem
17570    Steem
17571    Steem
17572    Steem
17573    Steem
17574    Steem
17575    Steem
17576    Steem
17577    Steem
17578    Steem
17579    Steem
17580    Steem
17581    Steem
17582    Steem
17583    Steem
17584    Steem
17585    Steem
17586    Steem
17587    Steem
17588    Steem
17589    Steem
17590    Steem
17591    Steem
17592    Steem
17593    Steem
17594    Steem
17595    Steem
17596    Steem
17597    Steem
         ...  
18162    Steem
18163    Steem
18164    Steem
18165    Steem
18166    Steem
18167    Steem
18168    Steem
18169    Steem
18170    Steem
18171    Steem
18172    Steem
18173    Steem
18174    Steem
18175    Steem
18176    Steem
18177    Steem
18178    Steem
18179    Steem
18180    Steem
18181    Steem
18182    Steem
18183    Steem
18184    Steem
18185    Steem
18186    Steem
18187    Steem
18188    Steem
18189    Steem
18190    Steem
18191    Steem
Name: name, Length: 624, dtype: object

## Competitors

In [2]:
some_symbols = ['BTC','ETH','DASH','XMR','QTUM','WAVES','LSK','OMG','STEEM','REP', 'GNT','SC','DGD','FCT','MAID','PAY','SNT','BAT','ZRX','BNT','CVC','STORJ','SINGLS','ANT','RLC','TRST','DNT']

blx_symbols = ['BTC', 'ETH','DASH','XMR','QTUM','WAVES','LSK','OMG','STEEM','REP','GNT','SC','DGD','FCT','MAID''PAY','SNT','BAT','ZRX','BNT','CVC','STORJ','SNGLS','ANT','RLC','TRST','DNT']

BLX: 

* Bitcoin (BTC)	Value	15.00 %	7.04 %
* Ethereum (ETH)	Platform	14.17 %	14.49 %
* Dash (DASH)	Anonymity	12.89 %	8.36 %
* Monero (XMR)	Anonymity	12.68 %	8.48 %
* Qtum (QTUM)	Platform	6.90 %	10.82 %
* Waves (WAVES)	Platform	6.50 %	3.22 %
* Lisk (LSK)	Platform	6.21 %	9.20 %
* OmiseGO (OMG)	Decentralized Exchange and Payments Platform	5.62 %	6.55 %
* Steem (STEEM)	Media	2.41 %	3.54 %
* Augur (REP)	Prediction	2.02 %	2.66 %
* Golem (GNT)	Decentralized Computing	1.75 %	2.04 %
* Siacoin (SC)	Storage	1.60 %	5.07 %
* DigixDAO (DGD)	Value Gold	1.47 %	1.22 %
* Factom (FCT)	Audit	1.33 %	1.52 %
* MaidSafeCoin (MAID)	Storage	1.27 %	1.19 %
* TenX (PAY)	Debit Card	1.27 %	1.04 %
* Status (SNT)	Smartphone Ethereum Gateway	1.20 %	3.56 %
* Basic Attention Token (BAT)	Advertisement	1.18 %	2.10 %
* 0x (ZRX)	Decentralized Exchange Protocol	0.86 %	2.90 %
* Bancor (BNT)	Financial Derivative	0.78 %	1.09 %
* Civic (CVC)	Identity Management	0.72 %	0.90 %
* Storj (STORJ)	Storage	0.54 %	0.72 %
* SingularDTV (SNGLS)	Media / TV	0.48 %	0.55 %
* Aragon (ANT)	Enterprise Management	0.43 %	0.54 %
* iExec (RLC)	Decentralized Computing	0.33 %	0.49 %
* WeTrust (TRST)	Lending	0.24 %	0.25 %
* District0x (DNT)	Decentralized infrastructure	0.15 %	0.46 %

### Comparison

Let's see what we have data for. 

In [47]:
valid_coins = []


def coinCheck(available_coins):
    valid_coins = []
    for ticker, symbol in available_coins:
            if len(df.loc[df.symbol == ticker].tail(365)):
                print('Last year of {} ({}) is available!'.format(ticker, name))
            else:
                print('Last year of {} ({}) is NOT available.'.format(ticker, name))
                
coinCheck(df_symbols)

ValueError: too many values to unpack (expected 2)

In [75]:
df.loc[df.symbol == 'QTUM'].tail(100)
['{} is available!'.format(name)
     if (len(df.loc[df.symbol == name].tail(365))>0) 
     else '{} is NOT available.'.format(name)
     for name in required_symbols]

slug symbol  name        date  ranknow   open   high    low  close  \
12348  qtum   QTUM  Qtum  2017-09-25       16   8.03   9.60   8.03   9.36   
12349  qtum   QTUM  Qtum  2017-09-26       16   9.38  10.30   9.17   9.34   
12350  qtum   QTUM  Qtum  2017-09-27       16   9.37   9.90   9.35   9.90   
12351  qtum   QTUM  Qtum  2017-09-28       16   9.95  10.10   9.06   9.93   
12352  qtum   QTUM  Qtum  2017-09-29       16   9.96  10.00   9.15   9.77   
12353  qtum   QTUM  Qtum  2017-09-30       16   9.77  11.65   9.75  11.59   
12354  qtum   QTUM  Qtum  2017-10-01       16  11.59  12.61  11.58  12.34   
12355  qtum   QTUM  Qtum  2017-10-02       16  12.36  12.38  11.43  11.88   
12356  qtum   QTUM  Qtum  2017-10-03       16  11.91  12.12  10.78  12.12   
12357  qtum   QTUM  Qtum  2017-10-04       16  12.16  12.28  11.51  11.63   
12358  qtum   QTUM  Qtum  2017-10-05       16  11.60  11.98  10.82  11.86   
12359  qtum   QTUM  Qtum  2017-10-06       16  11.87  11.91  11.50  11.69   
12360  qtum   QTUM  Qtum  2017-10-07       16  11.67  11.94  11.51  11.94   
12361  qtum   QTUM  Qtum  2017-10-08       16  11.96  12.11  11.14  11.14   
12362  qtum   QTUM  Qtum  2017-10-09       16  11.14  12.49  10.96  12.07   
12363  qtum   QTUM  Qtum  2017-10-10       16  12.04  12.54  11.55  12.08   
12364  qtum   QTUM  Qtum  2017-10-11       16  12.04  12.29  11.87  12.13   
12365  qtum   QTUM  Qtum  2017-10-12       16  12.13  12.21  11.29  11.29   
12366  qtum   QTUM  Qtum  2017-10-13       16  11.30  11.30  10.31  11.25   
12367  qtum   QTUM  Qtum  2017-10-14       16  11.26  11.45  10.62  10.63   
12368  qtum   QTUM  Qtum  2017-10-15       16  10.64  10.66  10.09  10.44   
12369  qtum   QTUM  Qtum  2017-10-16       16  10.46  11.65  10.23  11.49   
12370  qtum   QTUM  Qtum  2017-10-17       16  11.48  12.27  11.03  12.05   
12371  qtum   QTUM  Qtum  2017-10-18       16  12.04  12.04  10.78  11.49   
12372  qtum   QTUM  Qtum  2017-10-19       16  11.46  12.22  11.35  12.07   
12373  qtum   QTUM  Qtum  2017-10-20       16  12.10  15.12  11.19  11.30   
12374  qtum   QTUM  Qtum  2017-10-21       16  11.25  11.35  10.50  10.91   
12375  qtum   QTUM  Qtum  2017-10-22       16  10.97  11.06  10.56  10.63   
12376  qtum   QTUM  Qtum  2017-10-23       16  10.59  10.60   9.75   9.96   
12377  qtum   QTUM  Qtum  2017-10-24       16   9.96  11.46   9.52  10.76   
...     ...    ...   ...         ...      ...    ...    ...    ...    ...   
12418  qtum   QTUM  Qtum  2017-12-04       16  13.15  14.00  12.68  13.62   
12419  qtum   QTUM  Qtum  2017-12-05       16  13.69  13.95  13.08  13.08   
12420  qtum   QTUM  Qtum  2017-12-06       16  13.12  13.14  11.86  11.96   
12421  qtum   QTUM  Qtum  2017-12-07       16  11.95  14.93  11.67  12.56   
12422  qtum   QTUM  Qtum  2017-12-08       16  12.53  13.06  11.99  12.79   
12423  qtum   QTUM  Qtum  2017-12-09       16  12.77  12.77  11.54  12.08   
12424  qtum   QTUM  Qtum  2017-12-10       16  12.09  12.09  10.74  11.05   
12425  qtum   QTUM  Qtum  2017-12-11       16  11.02  14.60  11.00  13.55   
12426  qtum   QTUM  Qtum  2017-12-12       16  13.57  15.12  13.28  14.76   
12427  qtum   QTUM  Qtum  2017-12-13       16  14.81  19.29  14.80  17.31   
12428  qtum   QTUM  Qtum  2017-12-14       16  17.16  20.63  17.02  20.48   
12429  qtum   QTUM  Qtum  2017-12-15       16  20.68  32.65  17.69  30.54   
12430  qtum   QTUM  Qtum  2017-12-16       16  30.15  31.53  28.48  29.00   
12431  qtum   QTUM  Qtum  2017-12-17       16  29.19  30.30  28.63  29.62   
12432  qtum   QTUM  Qtum  2017-12-18       16  29.66  56.06  27.84  51.85   
12433  qtum   QTUM  Qtum  2017-12-19       16  51.96  77.75  51.30  67.75   
12434  qtum   QTUM  Qtum  2017-12-20       16  67.72  69.07  57.71  63.64   
12435  qtum   QTUM  Qtum  2017-12-21       16  63.82  69.24  56.87  62.23   
12436  qtum   QTUM  Qtum  2017-12-22       16  63.27  63.27  36.35  49.48   
12437  qtum   QTUM  Qtum  2017-12-23       16  50.84  60.20  50.71  57.26   
124

In [43]:
daily = pd.date_range('1/3/2016', end='1/3/2018', freq='D')
weekly = pd.date_range('1/3/2016', end='1/3/2018', freq='W')
print(daily)
print(weekly)

DatetimeIndex(['2016-01-03', '2016-01-04', '2016-01-05', '2016-01-06',
               '2016-01-07', '2016-01-08', '2016-01-09', '2016-01-10',
               '2016-01-11', '2016-01-12',
               ...
               '2017-12-25', '2017-12-26', '2017-12-27', '2017-12-28',
               '2017-12-29', '2017-12-30', '2017-12-31', '2018-01-01',
               '2018-01-02', '2018-01-03'],
              dtype='datetime64[ns]', length=732, freq='D')
DatetimeIndex(['2016-01-03', '2016-01-10', '2016-01-17', '2016-01-24',
               '2016-01-31', '2016-02-07', '2016-02-14', '2016-02-21',
               '2016-02-28', '2016-03-06',
               ...
               '2017-10-29', '2017-11-05', '2017-11-12', '2017-11-19',
               '2017-11-26', '2017-12-03', '2017-12-10', '2017-12-17',
               '2017-12-24', '2017-12-31'],
              dtype='datetime64[ns]', length=105, freq='W-SUN')


In [47]:
opt.matrix(np.cov(eth['open'], btc['open']))

<2x2 matrix, tc='d'>

In [33]:
np.asmatrix(eth['open'])

matrix([[  2.83000000e+00,   2.79000000e+00,   7.06136000e-01,
           7.13989000e-01,   7.08087000e-01,   1.06000000e+00,
           1.22000000e+00,   1.81000000e+00,   1.80000000e+00,
           1.68000000e+00,   1.58000000e+00,   1.22000000e+00,
           1.17000000e+00,   1.25000000e+00,   1.48000000e+00,
           1.40000000e+00,   1.38000000e+00,   1.35000000e+00,
           1.23000000e+00,   1.13000000e+00,   1.17000000e+00,
           1.15000000e+00,   1.19000000e+00,   1.18000000e+00,
           1.32000000e+00,   1.35000000e+00,   1.35000000e+00,
           1.30000000e+00,   1.26000000e+00,   1.28000000e+00,
           1.34000000e+00,   1.30000000e+00,   1.25000000e+00,
           1.24000000e+00,   1.21000000e+00,   1.16000000e+00,
           9.87228000e-01,   1.04000000e+00,   9.40566000e-01,
           8.75189000e-01,   9.41977000e-01,   9.06865000e-01,
           8.74574000e-01,   8.49603000e-01,   8.81191000e-01,
           9.41496000e-01,   9.19337000e-01,   9.035600

In [36]:
## NUMBER OF ASSETS
n_assets = 4

## NUMBER OF OBSERVATIONS
n_obs = 1000

return_vec = np.random.randn(n_assets, n_obs)

In [38]:
def rand_weights(n):
    ''' Produces n random weights that sum to 1 '''
    k = np.random.rand(n)
    return k / sum(k)

def random_portfolio(returns):
    ''' 
    Returns the mean and standard deviation of returns for a random portfolio
    '''

    p = np.asmatrix(np.mean(returns, axis=1))
    w = np.asmatrix(rand_weights(returns.shape[0]))
    C = np.asmatrix(np.cov(returns))
    
    mu = w * p.T
    sigma = np.sqrt(w * C * w.T)
    
    # This recursion reduces outliers to keep plots pretty
    if sigma > 2:
        return random_portfolio(returns)
    return mu, sigma

random_portfolio(return_vec)

(matrix([[ 0.00948996]]), matrix([[ 0.51000164]]))

### Todo:
* Pick the crypto to put in the portfolio (maybe top 20 by market cap in the data?)
* Generate a weight vector every week in the data, simulating real life (trained on the past year to date? All historical to current? Last week of data only?) 
* Given this weight vector, simulate a portfolio of that position for the week, and calculate expected performance.